src/Features/Core/Portable/AddParameter/AbstractAddParameterCodeFixProvider.cs

```diff
@@ -111,74 +154,164 @@ internal abstract class AbstractAddParameterCodeFixProvider<
-                        semanticModel, syntaxFacts, comparer, constructor,
-                        arguments, argumentOpt);
+                    var argumentToAdd = GetArgumentInsertPositionForMethodCandidate(semanticModel, syntaxFacts, method, arguments, argumentOpt);
+                    var typeArgumentToAdd = GetTypeArgumentForMethodCandidate(method, typeArguments);
 
-                    if (argumentToAdd != null)
+                    if (argumentToAdd != null || typeArgumentToAdd != null)
                     {
-                        if (argumentOpt != null && argumentToAdd != argumentOpt)
-                        {
-                            // We were trying to fix a specific argument, but the argument we want
-                            // to fix is something different.  That means there was an error earlier
-                            // than this argument.  Which means we're looking at a non-viable 
-                            // constructor.  Skip this one.
-                            continue;
-                        }
-
-                        constructorsAndArgumentToAdd.Add(
-                            (constructor, argumentToAdd, arguments.IndexOf(argumentToAdd)));
+                        methodsAndArgumentToAdd.Add(
+                            (method, argumentToAdd, arguments.IndexOf(argumentToAdd), typeArgumentToAdd));
                     }
                 }
             }
 
-            // Order by the furthest argument index to the nearest argument index.  The ones with
-            // larger argument indexes mean that we matched more earlier arguments (and thus are
-            // likely to be the correct match).
-            foreach (var tuple in constructorsAndArgumentToAdd.OrderByDescending(t => t.index))
+            return methodsAndArgumentToAdd.ToImmutableArray();
+        }
+
+        private TTypeSyntax GetTypeArgumentForMethodCandidate(IMethodSymbol method, SeparatedSyntaxList<TTypeSyntax> typeArguments)
+        {
+            var typeParameter = method.IsConstructor()
+                ? method.ContainingType.TypeParameters
+                : method.TypeParameters;
+
+            if (typeArguments.Count > typeParameter.Length)
             {
-                var constructor = tuple.constructor;
-                var argumentToAdd = tuple.argument;
+                return typeArguments[typeParameter.Length];
+            }
 
-                var parameters = constructor.Parameters.Select(p => p.ToDisplayString(SimpleFormat));
-                var signature = $"{type.Name}({string.Join(", ", parameters)})";
+            return null;
+        }
 
-                var title = string.Format(FeaturesResources.Add_parameter_to_0, signature);
+        private TArgumentSyntax GetArgumentInsertPositionForMethodCandidate(SemanticModel semanticModel, ISyntaxFactsService syntaxFacts, IMethodSymbol method, SeparatedSyntaxList<TArgumentSyntax> arguments, TArgumentSyntax argumentOpt)
```

Comment:
it's weird that this says "GetArgumentInsertPosition" but then ends up returning a TArgument.  Is htis returning the argument that the new argument will be added before/after?  if so, we should name things accordingly.